File: app/bot_scope.py
"""The default scope for a new bot: which channels, and which rooms.

A bot answers commands, so the conversations it listens to are the conversations
it can annoy. Left at "everywhere", enabling one built-in makes the node reply on
Public, on every other channel it happens to carry, and in every room server it
is logged into. A fresh bot therefore starts scoped to the two conventional bot
channels (``#bot`` / ``#bots``) plus DMs, and to no room at all; the operator
widens it from the Settings tab.

Hashtag keys are derived from the name, identical on every node, so the channel
default names channels this node has not joined yet: joining ``#bot`` later
brings the bot to life there with no scope edit. Rooms have no such convention —
no ``#bot`` room to fall back on, and a room answer is public to everyone logged
in — so they start as an empty pick list and the operator names the rooms a bot
may speak in.
"""

from __future__ import annotations

from typing import Any

from app.channel_constants import BOT_CHANNEL_KEYS

# The literal the ``bots.scope`` column defaults to (see app/database.py). Kept
# next to the builder so the two are checked against each other by the tests.
DEFAULT_BOT_SCOPE_JSON = (
    '{"channels": {"only": ["EB50A1BCB3E4E5D7BF69A57C9DADA211", '
    '"0D24F5830B449668B8C221759B6C50D2"]}, "rooms": {"only": []}}'
)


def no_rooms() -> dict[str, Any]:
    """The "answer in no room" selection: an empty pick list.

    Spelled ``{"only": []}`` rather than ``"none"`` so the editor opens on
    "Only…" with nothing ticked — the operator sees a list to add rooms to, not a
    switch to flip. Also what the engine falls back to for a scope that never
    named rooms at all, so rooms are opt-in however the scope was written.
    """
    return {"only": []}


def default_bot_scope() -> dict[str, Any]:
    """A fresh default-scope dict (never a shared mutable constant)."""
    return {"channels": {"only": list(BOT_CHANNEL_KEYS)}, "rooms": no_rooms()}


def is_default_bot_scope(scope: Any) -> bool:
    """True when ``scope`` selects exactly the default bot channels."""
    if not isinstance(scope, dict):
        return False
    channels = scope.get("channels")
    if not isinstance(channels, dict):
        return False
    only = channels.get("only")
    if not isinstance(only, list) or "except" in channels:
        return False
    return {str(key).upper() for key in only} == set(BOT_CHANNEL_KEYS)

File: frontend/src/utils/botScope.ts
import { CONTACT_TYPE_ROOM, type Channel, type Contact } from '../types';

/**
 * The channels a bot listens to out of the box: the two conventional bot
 * channels plus DMs. Bots answer commands, so an unscoped bot replies on Public
 * and on every other channel the node carries — noise for everyone else on the
 * mesh. The operator widens the scope from the bot's Settings tab.
 *
 * Hashtag keys are SHA256(name)[:16] of the verbatim name, so these are the
 * same keys on every node and stay meaningful even before this node joins the
 * channel. Mirrors `BOT_CHANNEL_KEYS` in `app/channel_constants.py`.
 */
export const DEFAULT_BOT_CHANNELS: { key: string; name: string }[] = [
  { key: 'EB50A1BCB3E4E5D7BF69A57C9DADA211', name: '#bot' },
  { key: '0D24F5830B449668B8C221759B6C50D2', name: '#bots' },
];

/** The name of a well-known default bot channel, even if we have not joined it. */
export function defaultBotChannelName(key: string): string | undefined {
  const upper = key.toUpperCase();
  return DEFAULT_BOT_CHANNELS.find((channel) => channel.key === upper)?.name;
}

/**
 * How to label a scoped channel key: the joined channel's name, else the
 * well-known bot-channel name, else a truncated key.
 */
export function scopeChannelLabel(key: string, channels: Channel[]): string {
  const joined = channels.find((channel) => channel.key.toUpperCase() === key.toUpperCase());
  return joined?.name ?? defaultBotChannelName(key) ?? `${key.slice(0, 8)}…`;
}

/** True when the key is not a channel this node has joined — the bot is deaf there. */
export function isUnjoinedChannel(key: string, channels: Channel[]): boolean {
  return !channels.some((channel) => channel.key.toUpperCase() === key.toUpperCase());
}

/** The room servers this node knows about, the order the pickers show them in. */
export function roomContacts(contacts: Contact[]): Contact[] {
  return contacts
    .filter((contact) => contact.type === CONTACT_TYPE_ROOM)
    .sort((a, b) => (a.name || a.public_key).localeCompare(b.name || b.public_key));
}

/**
 * How to label a scoped room key: the known room's name, else a truncated key.
 *
 * Unlike a hashtag channel there is no name to derive — a room is a contact, so
 * a key we have never heard an advert from can only be shown as itself.
 */
export function scopeRoomLabel(key: string, contacts: Contact[]): string {
  const known = contacts.find((contact) => contact.public_key.toLowerCase() === key.toLowerCase());
  return known?.name || `${key.slice(0, 8)}…`;
}

/** True when the key is not a room contact this node knows — the bot is deaf there. */
export function isUnknownRoom(key: string, contacts: Contact[]): boolean {
  return !roomContacts(contacts).some(
    (contact) => contact.public_key.toLowerCase() === key.toLowerCase()
  );
}
